### DISCUSSION TITLE
cleanup() is not triggered when <template x-for> element disappears

### DISCUSSION THREAD

**Discussion by harrynikolov** [alpinejs/alpine#3194] [7. Bugs (archived)]
I created a custom directive `x-countdown` which starts a 1-second interval timer. In the cleanup() function, I want to remove this timer.

This directive is nested deep inside a for-loop, which looks like this: `<template x-for="..." :key="...">`

When the array changes, the items change correctly. However, the custom directive's `cleanup()` function is not called. (I have confirmed this by placing `console.log` across the entire directive)

**UPDATE:**
It looks like the `x-data` in-between is what's causing this issue. _Any_ value inside `x-data` will cause this, doesn't matter what's inside x-data, only its presence matters.

**Comment by ekwoka** [comment]
Can you show the directive?

**Reply by harrynikolov** [reply]
Sure, hope this helps. Here is the directive's source code:
```
document.addEventListener('alpine:init', () => {
    Alpine.directive('countdown', (el, { expression }, { evaluate, cleanup }) => {
        let seconds = evaluate(expression);

        const tick = () => {
            let d = Math.floor(seconds / (60 * 60 * 24));
            let h = Math.floor((seconds % (60 * 60 * 24)) / (60 * 60));
            let m = Math.floor((seconds % (60 * 60)) / (60));
            let s = Math.floor((seconds % 60));
            el.innerHTML = `${d}d ${h}h ${m}m ${s}s`;

            if (seconds === 0) {
                clearInterval(interval);
                console.log("NATURAL CLEANUP");
            } else {
                seconds--;
            }
        }

        tick();
        const interval = setInterval(tick, 1000);

        cleanup(() => {
            clearInterval(interval);
            console.log("FORCED CLEANUP");
        })
    });
});
```

And here is how it's used:
```
<template x-for="package in $store.data.packages" :key="package.id">
    <div class="package" x-data="packageControls(package)">

        <template x-if="package.expirySeconds">
            <div>
                EXPIRING SOON:
                <span x-countdown="package.expirySeconds"></span>
            </div>
        </template>

        ...

    </div>
</template>
```

**Comment by harrynikolov** [comment]
**UPDATE:**
It looks like the `x-data` in-between is what's causing this issue. _Any_ value inside x-data will cause this, doesn't matter what's inside x-data, only its presence matters.
